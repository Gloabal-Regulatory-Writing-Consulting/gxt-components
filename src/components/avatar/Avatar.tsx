import React, { ChangeEvent, useMemo } from "react";
import { IconType } from "../svg/SvgIcon";
import {
  AvatarBox,
  AvatarButton,
  AvatarContainer,
  AvatarIcon,
  AvatarText,
  StyledIcon,
  UserIcon,
  UserInitials,
} from "./AvatarStyledComponents";

export type AvatarProps = {
  size: "small" | "large";
  onClickHandler?: () => void;
  handleImageUpload?: (e: ChangeEvent<HTMLInputElement>) => void;
  user: UserObj;
  isExpanded?: boolean;
  imageUrl?: string;
};

export type UserObj = {
  firstName: string;
  lastName: string;
  avatar?: string;
};

const Avatar: React.FC<AvatarProps> = ({
  size,
  onClickHandler,
  handleImageUpload,
  user,
  isExpanded,
  imageUrl,
}: AvatarProps) => {
  const { firstName, lastName, avatar } = user;
  const userName = `${firstName} ${lastName}`;

  const userInitials = useMemo(() => {
    const lastNameParts = lastName?.split(/[ -]/).filter((e) => e);
    const lastPart = lastNameParts?.at(-1) ?? "";
    return (firstName?.charAt(0) || "") + (lastPart?.charAt(0) || "");
  }, [firstName, lastName]);

  const noIcon: JSX.Element = (
    <UserInitials data-testid="user-initials" size={size}>{userInitials}</UserInitials>
  );

  const userIcon: JSX.Element = (
    <UserIcon src={imageUrl || avatar} alt="Avatar" />
  );

  const userAvatar = useMemo(() => {
    switch (true) {
        case !!imageUrl || !!avatar:
            return userIcon;
        default:
            return noIcon;
    }
}, [imageUrl, avatar, userIcon, noIcon]);

  const editImage = handleImageUpload ? (
    <>
      <StyledIcon
        iconType={IconType.PENCIL}
        size={size === "small" ? 8 : 18}
      ></StyledIcon>
      <input
        type="file"
        data-testid="user-image-input"
        style={{ display: "none" }}
        id="img"
        name="img"
        accept="image/*"
        onChange={handleImageUpload}
      />
    </>
  ) : null;

  return (
    <AvatarContainer className={`${isExpanded && "expanded-avatar"}`}>
      <AvatarIcon data-testid="avatarIcon" onClick={onClickHandler}>
        <AvatarBox>
          <AvatarButton size={size} className={`avatar-button`}>
            {userAvatar}
          </AvatarButton>
          {<label htmlFor="img"> {editImage} </label>}
        </AvatarBox>
        {isExpanded && (
          <div>
            <AvatarText weight="bold" size={size}>
              {userName}
            </AvatarText>
            <AvatarText weight="normal" size={size}>View profile</AvatarText>
          </div>
        )}
      </AvatarIcon>
    </AvatarContainer>
  );
};

export default Avatar;
