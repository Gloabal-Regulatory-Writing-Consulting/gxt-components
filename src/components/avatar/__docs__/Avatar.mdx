import { Canvas, Meta, Controls } from "@storybook/blocks";
import * as Avatar from "./Avatar.stories.tsx";

<Meta of={Avatar} title="Avatar" />

# Avatar

Avatar component with different props.

#### Example

<Canvas of={Avatar.AvatarWithUpload} />

## Usage

```tsx
import Avatar from "sld-ui";

const Example = () => {
  const onClickHandler = () => {
    // Handle on click of avatar
  };

  const handleImageUpload = (e) => {
    // Handle image upload
  };

  return (
    <Avatar
      size="small"
      user={{ firstName: "John", lastName: "Doe", avatar: "avatar.jpg" }}
      isExpanded={false}
      imageUrl="image.jpg"
      onClickHandler={onClickHandler}
      handleImageUpload={handleImageUpload}
    />
  );
};

export default Example;
```

## Props
<Controls of={Avatar.AvatarWithUpload} />
