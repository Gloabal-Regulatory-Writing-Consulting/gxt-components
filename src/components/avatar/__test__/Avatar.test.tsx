import React, { act } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Avatar, { AvatarProps, UserObj } from '../Avatar';
import { describe, expect, it, vi } from 'vitest';

describe('Avatar', () => {
  const mockUser: UserObj = {
    firstName: 'John',
    lastName: 'Doe',
  };

  const renderComponent = (props?: Partial<AvatarProps>) => {
    const defaultProps: AvatarProps = {
      size: 'small',
      user: mockUser,
    };
    return render(<Avatar {...defaultProps} {...props} />);
  };

  it('renders the user initials when no image is provided', () => {
    renderComponent({ imageUrl: undefined });

    const userInitials = screen.getByTestId('user-initials');
    expect(userInitials).toBeInTheDocument();
    expect(userInitials.textContent).toBe('JD');
  });

  it('renders the user image when imageUrl is provided', () => {
    renderComponent({ imageUrl: 'image.jpg' });

    const userImage = screen.getByAltText('Avatar');
    expect(userImage).toBeInTheDocument();
    expect(userImage).toHaveAttribute('src', 'image.jpg');
  });

  it('calls onClickHandler when the avatar button is clicked', () => {
    const onClickHandler = vi.fn();
    renderComponent({ onClickHandler });

    const avatarButton = screen.getByTestId('avatarIcon');
    fireEvent.click(avatarButton);

    expect(onClickHandler).toHaveBeenCalledTimes(1);
  });

  it('calls handleImageUpload when a file is selected', () => {
    const handleImageUpload = vi.fn();
    renderComponent({ handleImageUpload });

    const fileInput = screen.getByTestId('user-image-input');
    fireEvent.change(fileInput, { target: { files: ['image.jpg'] } });

    expect(handleImageUpload).toHaveBeenCalledTimes(1);
  });

  it('renders the user name and "View profile" text when isExpanded is true', () => {
    renderComponent({ isExpanded: true });
    act(() => {
      const userName = screen.getByText("John Doe");
      const viewProfileText = screen.getByText("View profile");
      expect(userName).toBeInTheDocument();
      expect(viewProfileText).toBeInTheDocument();
    });
  });
});
