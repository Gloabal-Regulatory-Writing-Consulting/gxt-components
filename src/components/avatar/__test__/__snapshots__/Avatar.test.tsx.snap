// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > Avatar should render with a specific size > AvatarLargeSize 1`] = `
<DocumentFragment>
  <div
    class="sc-aYaIB iAVyJI undefined"
  >
    <div
      class="sc-gEvDqW KVfas"
      data-testid="avatarIcon"
    >
      <div
        class="sc-eqUzNf engioi"
      >
        <div
          class="sc-fqkwJk eUDqXp avatar-button"
        >
          <div
            class="sc-iGgVNO gqEbas"
            data-testid="user-initials"
          >
            JD
          </div>
        </div>
        <label
          for="img"
        >
            
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > Avatar should render with a specific size > AvatarLargeSize 2`] = `
<DocumentFragment>
  <div
    class="sc-aYaIB iAVyJI undefined"
  >
    <div
      class="sc-gEvDqW KVfas"
      data-testid="avatarIcon"
    >
      <div
        class="sc-eqUzNf engioi"
      >
        <div
          class="sc-fqkwJk ffocQB avatar-button"
        >
          <div
            class="sc-iGgVNO fnVQo"
            data-testid="user-initials"
          >
            JD
          </div>
        </div>
        <label
          for="img"
        >
            
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
